implementations/micrometer-registry-cloudwatch/src/main/java/io/micrometer/cloudwatch/CloudWatchMeterRegistry.java

```diff
@@ -0,0 +1,213 @@
+/**
+ * Copyright 2017 Pivotal Software, Inc.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.cloudwatch;
+
+import com.amazonaws.handlers.AsyncHandler;
+import com.amazonaws.services.cloudwatch.AmazonCloudWatchAsync;
+import com.amazonaws.services.cloudwatch.model.Dimension;
+import com.amazonaws.services.cloudwatch.model.MetricDatum;
+import com.amazonaws.services.cloudwatch.model.PutMetricDataRequest;
+import com.amazonaws.services.cloudwatch.model.PutMetricDataResult;
+import com.amazonaws.services.cloudwatch.model.StandardUnit;
+import io.micrometer.core.instrument.Clock;
+import io.micrometer.core.instrument.DistributionSummary;
+import io.micrometer.core.instrument.FunctionTimer;
+import io.micrometer.core.instrument.HistogramSnapshot;
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Timer;
+import io.micrometer.core.instrument.ValueAtPercentile;
+import io.micrometer.core.instrument.config.NamingConvention;
+import io.micrometer.core.instrument.histogram.HistogramConfig;
+import io.micrometer.core.instrument.step.StepMeterRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.text.DecimalFormat;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.StreamSupport.stream;
+
+/**
+ * @author Dawid Kublik
+ */
+public class CloudWatchMeterRegistry extends StepMeterRegistry {
```

Comment:
Why did you choose to extend `StepMeterRegistry` instead of just `MeterRegistry`?

Reply:
I started with MeterRegistry and CumulativeCounter but then ClouWatch counter metrics doesn't behave as I would expect to.

With CloudWatch newly sent metric is being aggregated into an existing metric so there is no access to original metrics but only to aggregated form (Avg, Min, Max, Sum, Sample count, percentiles), with default one minute resolution.


eg. test: every 5 seconds I increase counter by 10,
expectation: 12 * 10 = 120 avg outcome for every minute,


with cumulative counter:

time		00	05	10	15	20	25	30	35	40	45	50	55	
counter 	10	20	30	40	50	60	70	80	90	100	110	120

e.g. sampling: 06, 26, 46
avg: (20 + 60 + 100) / 3 = 60
min: 20
max: 100
sum: 180

none of these is of any usage and all of them just go up and up until everything is messeed up even more with server restart.


with step counter:

time		00	05	10	15	20	25	30	35	40	45	50	55	
counter 	10	20	10	20	30	40	10	20	30	40	10	20

e.g. sampling: 06, 26, 46
avg: (20 + 40 + 40) / 3 = 33.3 (40 every minute after first)
min: 20 (40 every minute after first)
max: 40
sum: 100 (120 every minute after first)

This can be generalized to other sampling periods and and non regular metrics sending.

similar things can be reason about Timer and StepTimer.


stepCounter Sum:
<img width="772" alt="stepcounter" src="https://user-images.githubusercontent.com/5848865/33214174-0fce9740-d12b-11e7-87a0-6901203a4cdb.png">

cumulativeCounter Avg:
<img width="805" alt="cumulativecounter" src="https://user-images.githubusercontent.com/5848865/33214177-1114f0d6-d12b-11e7-906b-1ea7f085f76a.png">

Reply:
I think you did the right thing here @dkublik. Most monitoring systems don't do arbitrary rate normalization on the backend. Step registries are really the norm, with Prometheus being a notable exception.

Reply:
Agreed! Thanks for you clear explanation and work on this.